### ISSUE TITLE
koreader incorrectly displays text in <section> tags

### ISSUE THREAD

**Issue by alethiophile** [koreader/koreader#3987]
* KOReader version: v2015.11-1654-g666c9f1_2018-04-12
* Device: Kobo Aura

#### Issue
When displaying an epub, any text directly under a \<section> tag fails to display. Text under descendent tags such as \<p>, \<i>, \<b> &c. displays correctly.

This is a problem when dealing with epubs that wrap their chapters in \<section> tags and typeset using \<br> rather than \<p>; in these files, most of the text is not displayed.

#### Steps to reproduce
1. Create an epub with text directly under a \<section> tag
2. View it

##### `crash.log` (if applicable)
There is no crash.

**Comment by poire-z** [comment]
In crengine, all these tags get flagged with `allow_text = false` `///< is true if element allows text nodes as children`. Meaning they can only contain other elements.
Is that an error for `section`? And are there other errors in that list?
```
XS_TAG1( FictionBook )
XS_TAG1( annotation )
XS_TAG1( output )
XS_TAG1( part )
XS_TAG1( param )
XS_TAG1( coverpage )
XS_TAG1( section )
XS_TAG1( svg )
XS_TAG1( dl )
XS_TAG1( ol )
XS_TAG1( ul )
XS_TAG1( html )
XS_TAG1( head )
XS_TAG1( hr )
XS_TAG1( stanza )
XS_TAG1( epigraph )
XS_TAG1( poem )
XS_TAG1( author )
XS_TAG1( DocFragment )
```

**Comment by poire-z** [comment]
It also looks like there is special treatment for `section` and that it may be of use related to crengine footnotes (that we don't use), or for building TOC items (search for `el_section` and `_isSection` in lvrend.cpp, lvtinydom.cpp, lvdocview.cpp)

**Comment by Frenzie** [comment]
The majority of those aren't HTML so I have no idea. But even on elements that *technically* don't allow text nodes directly (mainly the list elements like `ol`, `ul`, and `dl`) I don't think it makes much sense from a HTML perspective to not to display whatever text they might contain.

**Comment by alethiophile** [comment]
I can't speak to anything epub-specific, but from what I can tell `<section>` should allow plain text underneath it in HTML. The calibre epub reader allows text under it, as does firefox in an HTML file.

**Comment by Frenzie** [comment]
EPUB is just XHTML, or in other words HTML with a stricter syntax. Those weird other elements are for FictionBook (FB2), except for `DocFragment` which is for internal use.

@poire-z Btw a `section` is pretty much the same thing as a `div`.

**Comment by cramoisi** [comment]
`<section>` tags have nothing to do in a well formed epub. They are probably inherit from a fb2 conversion. (all sections in fb2 are chapters). As calibre epub reader is a lot permissive with nearly everything I wouldn't take the fact it displays it as a fact. It's probably calibre which didn't convert the tag in the first place.

**Comment by cramoisi** [comment]
saying that, section is a valid tag in html 5 with more or less the same intend than the one in fb2, so ..? (even if I don't see what it could do a div tag couldn't)

**Comment by Frenzie** [comment]
Div is more generic but I agree that section feels somewhat redundant.

**Comment by Frenzie** [comment]
From your link:

> It can be used within articles, in the header or footer, or to define navigation.

That sounds patently wrong due to the existence of nav. >_>

**Comment by alethiophile** [comment]
Leaving aside what purpose a `<section>` tag serves in an epub, the fact remains that I have a lot of epubs with all their text inside them, and it would be nice to be able to read them.

"Generous in what you accept" seems like it should apply here.

**Comment by cramoisi** [comment]
@alethiophile yup, anyway, if it's there and it's valid, koreader should be able to display it. After a little ressearch I noticed InDesign encourages designers to use some of [these specifics tags in place of generic one (span, div)](http://epubsecrets.com/avoiding-generic-html.php). So I guess I would be wise to make them work.

**Comment by Frenzie** [comment]
They're quite right of course. :-)

**Comment by cramoisi** [comment]
and, I should say (before going to bed) this will be quite common [with epub3 being more and more use](https://www.edrlab.org/epub/epub2-vs-epub3/).

I know calibre now can convert in epub3 but i didn't have the time to look at what it does when it does it.

**Comment by poire-z** [comment]
I guess it's just a matter of changing XS_TAG1( element ) to XS_TAG1T( element ) (and ensuring backward fallback to allow_text=false according with dom_version):
<kbd>![image](https://user-images.githubusercontent.com/24273478/40629913-85d849ce-62ce-11e8-8061-5cccf9507aa2.png)</kbd>.
The other special treatments of el_section seem they can live their life either way.
After all, if the FB2 specs say SECTION shouldn't have any text content, well, FB2s documents don't have them and we won't be displaying something that's not there :) If HTML have some SECTION with text, they'll be displayed.

So, among the above list, we should use XS_TAG1T for SECTION, OL, UL, DL ? Any other ?
(HEAD, HTML, SVG, HR, PARAM (stuff used in flash/activeX object, no?) should stay XS_TAG1 I guess).

While we're at it, and now that all unknown elements are `display: inline`, it would be nice to have a list of what the latest specs define as being `display: block`, to include them in fb2def.h or in epub.css. Any one up to making a list?
cf the list of the known elements https://github.com/koreader/crengine/blob/master/crengine/include/fb2def.h#L30-L156 and last week discussion at https://github.com/koreader/crengine/pull/191#issuecomment-391653311

**Comment by Frenzie** [comment]
> HEAD

In HTML that would signal the start of the BODY element, so in that sense yes. But not strictly speaking for HEAD.

HR (nor any self-closing element) cannot contain text but in principle if someone wrote `<hr>bla</hr>` it should probably be interpreted as `<hr>bla`. The HTML5 spec details that kind of error handling somewhere though it's not really the biggest concern I think? :-) PARAM is another one of those.

SVG I don't think so but isn't that for the SVG renderer to worry about anyway?

> it would be nice to have a list of what the latest specs define as being display: block, to include them in fb2def.h or in epub.css

I would assume this is accurate, though I'm not sure if it's complete: https://developer.mozilla.org/en-US/docs/Web/HTML/Block-level_elements

Where by complete I mean in the sense of old deprecated elements that probably shouldn't matter anyway. ;-)

**Comment by poire-z** [comment]
Quick sorting of those elements on the MDN page.
Those that we don't have, and that are like `<div>` with semantic:
```
address
article
aside
figcaption
figure
footer
header
hgroup
main
nav
section (we have this one, to fix to show text nodes)
```
Others with specific use:
```
canvas
fieldset // usually related to FORM
noscript // why not as we obviously don't support script
output  // usually related to FORM
video
```
We have this, that simply hide all `<form>` content (which is good, as we have no code to draw any form input element):
`XS_TAG1D( form, true, css_d_none, css_ws_normal ) /* not displayed */`

> SVG I don't think so but isn't that for the SVG renderer to worry about anyway?

We only render external SVG image. We have nothing to deal with embedded SVG XML in a HTML page. I guess such embedded SVG text nodes would be displayed (if SVG have text nodes, dunno), which I guess we should not as it would be just crap.

**Comment by Frenzie** [comment]
> We have this, that simply hide all <form> content (which is good, as we have no code to draw any form input element):

Don't we? `input{border:1px solid black}` → input element drawn, just won't receive any input

> We only render external SVG image. We have nothing to deal with embedded SVG XML in a HTML page. I guess such embedded SVG text nodes would be displayed (if SVG have text nodes, dunno), which I guess we should not as it would be just crap.

Oh, that way. You would normally expect any text inside elements to render as fallback content, although to what extent that may or may not be desirable with SVG is hard to say. [This](https://developer.mozilla.org/en-US/docs/Web/SVG/Element/text) example actually renders very nicely in Links2:

![screenshot_2018-05-29_14-12-03](https://user-images.githubusercontent.com/202757/40658072-4c260c12-634a-11e8-943f-ec32d6ef1f42.png)

**Comment by poire-z** [comment]
> Don't we? input{border:1px solid black} → input element drawn, just won't receive any input

Guess that's just an unknown element that you disguised with style... If included in a FORM, it should be hidden.

What about `canvas fieldset noscript output video `?

**Comment by Frenzie** [comment]
Relevant selection of the concrete CSS I used with Prince:[1]

```css
fieldset {margin-bottom:1em; background: #eee;border:black solid thin;border-radius:1em}
legend {padding: .5em; margin-left:.5em;font-weight: bold; background: lightgreen; border: black solid thin;border-radius:1em}
body {counter-reset: section}
legend::before { counter-increment:section; content: counter(section) ". "; }
fieldset:last-child td {width:50%; font-size: .8em; text-align: left}

input, textarea {background: #fff; border:black solid thin}
textarea {display:block;width:60%;height:6em}

@media print {
	html {font-size: 9.5pt}
	fieldset {page-break-inside:avoid}
	/* for Prince */
	input {display:inline-block;vertical-align:top;content: attr(value) " "; width: 11em;height:1.5em}
	input[size="2"]{width: 2em}
	input[size="3"]{width: 3em}
	input[size="4"]{width: 4em}
	input[size="6"]{width: 6em}
	input[size="18"]{width: 18em}
	input[type="checkbox"]{width:1.5em}
	@page { size: A4 }
	@page {margin:4em 2em 2em}
	@page {
		@top { content: flow(header) }
		@bottom {content: "Pagina " counter(page) " van " counter(pages)}
	}
	h1 { flow: static(header) }
}
```

Prince rendered result:
![screenshot_2018-05-29_17-21-04](https://user-images.githubusercontent.com/202757/40668407-b1848fa6-6364-11e8-91f7-4c47f168fced.png)

[ned-taalbeheersing-2-formulier.pdf](https://github.com/koreader/koreader/files/2048852/ned-taalbeheersing-2-formulier.pdf)

[1] Slightly messy. Presumably I threw it together in a bit of a hurry. My standard actually looks more like `fieldset {margin-bottom:1em; background:#eee; border:black solid thin; border-radius:1em}`. I find it much easier to read than the long drawn-out stuff, but unfortunately changing a property changes the whole line, so for version control it's not great. Also many other people don't like it that way. :-P

**Comment by Frenzie** [comment]
`fieldset` and `form` itself are block, other than that I can't think of any otoh. :-P

**Comment by poire-z** [comment]
Should be fixed by #3992, which is in current nightly and release.
